src/smt/optimization_solver.h

```diff
@@ -80,38 +90,350 @@ class Objective
-   * spawns a subsolver SmtEngine that captures the parent assertions and 
-   * implements a linear optimization loop. Supports activateObjective, 
-   * checkOpt, and objectiveGetValue in that order.
-   */
-  class OptimizationSolver
-  {
-   public:
-    /** parent is the smt_solver that the user added their assertions to **/
-    OptimizationSolver(SmtEngine* parent);
-    ~OptimizationSolver();
-
-    /** Runs the optimization loop for the activated objective **/
-    OptResult checkOpt();
-    /** Activates an objective: will be optimized for **/
-    void activateObj(const Node& obj, const int& type);
-    /** Gets the value of the optimized objective after checkopt is called **/
-    Node objectiveGetValue();
-
-   private:
-    /** The parent SMT engine **/
-    SmtEngine* d_parent;
-    /** The objectives to optimize for **/
-    Objective d_activatedObjective;
-    /** A saved value of the objective from the last sat call. **/
-    Node d_savedValue;
-  };
+
+  /** Specify whether to use signed or unsigned comparison
+   * for BitVectors (only for BitVectors), this variable is defaulted to true
+   * **/
+  bool d_signed;
+};
+
+
+/**
+ * A solver for optimization queries.
+ *
+ * This class is responsible for responding to optmization queries. It
+ * spawns a subsolver SmtEngine that captures the parent assertions and
+ * implements a linear optimization loop. Supports activateObjective,
+ * checkOpt, and objectiveGetValue in that order.
+ */
+class OptimizationSolver
+{
+ public:
+  /** parent is the smt_solver that the user added their assertions to **/
+  OptimizationSolver(SmtEngine* parent);
+  ~OptimizationSolver();
+
+  /** Runs the optimization loop for the activated objective **/
+  OptResult checkOpt();
+  /** Activates an objective: will be optimized for
+   * Parameter is_signed specifies whether we should use signed/unsigned
+   * comparison for BitVectors (only effective for BitVectors) **/
+  void activateObj(const Node& obj,
+                   const int& type,
+                   bool bv_is_signed_compare = true);
+  /** Gets the value of the optimized objective after checkopt is called **/
+  Node objectiveGetValue();
+
+//  private:
+//   /** Returns the less than operator for the activated objective
+//    * if objective does not support comparison, return kind::NULL_EXPR **/
+//   Kind getLessThanOperatorForObjective();
+
+ private:
+  /** The parent SMT engine **/
+  SmtEngine* d_parent;
+  /** The objectives to optimize for **/
+  Objective d_activatedObjective;
+  /** A saved value of the objective from the last sat call. **/
+  Node d_savedValue;
+};
+
+/**
+ * Optimizer for individual datatype 
+ */
+struct OMTOptimizer 
+{
+  static std::unique_ptr<OMTOptimizer> getOptimizerForNode(CVC4::Node node, bool isSigned=false);
+  virtual ~OMTOptimizer() {}
+  virtual std::pair<OptResult, CVC4::Node> minimize(SmtEngine *parentSMTSolver, CVC4::Node target) = 0;
```

Comment:
I think it would make sense for the pattern that we use it to be similar to the API. In the API draft we did not return pairs. There, we have two getters, when for the result and one for the term (Term in the API is Node here).
https://github.com/mcjuneho/CVC4/blob/a5d32ad5afa115386d53ca7c17eb2d7869cdb565/src/api/cvc4cpp.h#L2258

Even not going far to the API, is there a reason not to follow the current pattern on master?
https://github.com/CVC4/CVC4/blob/6beb70fcedd18e965ad82949090365cb44a43692/src/smt/optimization_solver.h#L93

Reply:
Disagreed. The original API is bad in this case -- maximize / minimize is supposed to return a <optimizationOutcome, optimizedValue> pair. 
The original API is just returning the outcome and letting the user ask for the value, which is completely nonsense -- the user is desperately begging for the optimization VALUE (not the inhumanity outcome saying the solver gets something sat/unsat/unbounded), and the optimization outcome is just to explain why the user gets such value.

Reply:
The idea behind the original api design was to be similar to `(check-sat)` and then `(get-value)` or `(get-model)`. I think it is good for the API to be consistent when possible, and getting the optimal value is pretty similar to just getting a value, right? Similarly at the level of the smt-engine, where check-sat and get-value calls are separated and not unified to return a pair.

BTW this very discussion is a good motivation to start with the API PR. These are issues that should first be resolved there and then propagated to more specific code like this one. The original API PR was obtained after several discussions, and we should continue these discussions if you think one of our decisions were not optimal. The right place to discuss this would be in an API PR, where u propose an alternative to the existing one.

Reply:
ok~ then let's make a PR to cvc4/master for this branch

Reply:
sure, that's the goal. What I am trying to do with the current PR is to clear initial feedback I expect we'll get in a PR for master. In particular, do u have any thoughts about the similarly to check-sat/get-value? Is there a reason for it to be different in an optimization setting? Another good way to think about it is to look at APIs of other solvers -- do they separate or return a pair?

Reply:
I would like to say it's nothing, because returning a pair doesn't affect the outer API because the optimizer is private. whatever it returns is not defined in the optimizer. But still, as a programmer if you want 2 values just return 2 values.

Reply:
That makes sense, the only thing remaining thing about this that I think we need to think about is consistency w.r.t. check-sat/get-value. Is there a good reason to use a pair for optimization but not to use a pair in check-sat/get-value?

Reply:
because that will introduce extra field in the OMTOptimizer base class. and it's not necessary.

Reply:
can't you say the same about get-value that requires an extra `model` (or something like that) member?

Reply:
YES but that's how the API defines.

Reply:
this one is not interacting with the API. so why introduce an extra field if you only need the result once

Reply:
I will also stop discussing this aesthetic thing, which is the same as below.